python/cog/server/clients.py

```diff
@@ -243,11 +248,14 @@ async def upload_file(
                 url = resp1.headers["Location"]
 
         log.info("doing real upload to %s", url)
-        resp = await self.file_client.put(
-            url,
-            content=ChunkFileReader(fh),
-            headers=headers,
-        )
+        if isinstance(fh, OutputURL):
+            async with self.file_client.stream("GET", fh.url) as resp:
+                content = resp.aiter_bytes()
+                resp = await self.file_client.put(url, content=content, headers=headers)
+        else:
+            resp = await self.file_client.put(
+                url, content=ChunkFileReader(fh), headers=headers
+            )
```

Comment:
Thanks for pulling this together. I'm trying to understand this, I'm not very familiar with how `async/await` works in Python. I understand by going with the `httpx` async client + iterator we get a non-blocking implementation but how does the existing `ChunkFileReader` accomplish the same thing?

If it doesn't and we still need to solve that issue we should probably figure out an interface that works for both.

Reply:
The PUT method takes bytes or an async iterator that returns bytes. aiter_bytes counts, as does ChunkFileReader, which is implemented earlier in the same file. ChunkFileReader does do blocking disk reads, but doing so 1MB at a time is likely short enough that we can do all the other networking we need in between.

if you wanted to be fancy you could have a single FileIterator that could take a local or remote URI, but that's kind of annoying to do while holding the context manager for the download request, and this approach is kind of simpler